Frameworks/CoreGraphics/CGImage.mm

```diff
@@ -390,594 +169,217 @@ CGImageRef CGImageMaskCreate(size_t width,
 
-        for (unsigned x = 0; x < width; x++) {
-            BYTE r = *rowIn++;
-            BYTE g = *rowIn++;
-            BYTE b = *rowIn++;
-            BYTE a = *rowIn++;
+    ComPtr<IWICBitmap> image;
+    ComPtr<IWICImagingFactory> imageFactory = _GetWICFactory();
 
-            BYTE alphaOut = (r + g + b) * a / 255 / 3;
-            *rowOut++ = 255 - alphaOut;
-        }
+    CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceGray();
+    REFGUID pixelFormat = _CGImageGetWICPixelFormat(bitsPerComponent, bitsPerPixel, colorSpace, kCGBitmapByteOrderDefault);
+    CGColorSpaceRelease(colorSpace);
 
-        pixOut += outImageBytesPerRow;
-    }
+    HRESULT status = imageFactory->CreateBitmapFromMemory(width, height, pixelFormat, bytesPerRow, height * bytesPerRow, data, &image);
 
-    newImage->Backing()->ReleaseImageData();
+    CGImageRef imageRef = __CGImage::CreateInstance();
+    imageRef->ImageSource().Attach(image.Detach());
+    imageRef->SetIsMask(true).SetInterpolate(shouldInterpolate);
 
-    return (CGImageRef)newImage;
+    return imageRef;
 }
 
 /**
  @Status Interoperable
 */
-CGImageAlphaInfo CGImageGetAlphaInfo(CGImageRef img) {
-    int32_t ret;
-
-    if (img) {
-        ret = img->Backing()->BitmapInfo() & kCGBitmapAlphaInfoMask;
-    } else {
-        TraceWarning(TAG, L"CGImageGetAlphaInfo: Null CGImageRef!");
-        ret = 0;
-    }
-
-    return (CGImageAlphaInfo)ret;
-}
-
-@interface CGImageDataProvider : NSData {
-@public
-    CGImageRef _img;
-}
+CGDataProviderRef CGImageGetDataProvider(CGImageRef img) {
+    RETURN_NULL_IF(!img);
 
-- (instancetype)init;
-- (instancetype)initWithBytesNoCopy:(void*)bytes length:(NSUInteger)length freeWhenDone:(BOOL)freeWhenDone;
-- (const void*)bytes;
-- (NSUInteger)length;
+    size_t height, width;
+    RETURN_NULL_IF_FAILED(img->ImageSource()->GetSize(&width, &height));
 
-@end
+    const unsigned int stride = width * CGImageGetBytesPerRow(img);
+    const unsigned int size = height * stride;
+    woc::unique_iw<unsigned char> data(static_cast<unsigned char*>(IwMalloc(size)));
 
-// TODO: Task 7188763 This class makes no sense to be derived from NSData as it exposes a public _img field
-// and apparently does all operations via that rather than actually acting like an NSData.
-// To make it work, just add in the appropriate NSData methods using an inner NSData to hold anything
-// with the assumption it is not used.
-@implementation CGImageDataProvider {
-@private
-    StrongId<NSData> _data;
-}
-- (void)dealloc {
-    _img->Backing()->ReleaseImageData();
-    [super dealloc];
-}
+    RETURN_NULL_IF_FAILED(img->ImageSource()->CopyPixels(0, stride, stride, data.get()));
 
-- (instancetype)init {
-    return [self initWithBytes:"" length:0];
+    NSData* byteData = [NSData dataWithBytesNoCopy:data.release() length:size freeWhenDone:YES];
+    CGDataProviderRef ret = CGDataProviderCreateWithCFData((CFDataRef)byteData);
+    CFAutorelease(ret);
+    return ret;
```

Comment:
Please don't use CFAutorelease if possible. The fact that you are using it here suggests that the dataProvider object maybe should be a member of CGImage instead of created everytime? This could also be a more efficient solution as a new pixel buffer wouldn't need to be allocated everytime. I'm not sure if / how the image can change but maybe consider creating on the first call and assigning a member variable that you return otherwise?